gc/2021/GC-10-05.md

```diff
@@ -0,0 +1,192 @@
+1. Find volunteers for note taking (acting chair to volunteer)
+1. Adoption of the agenda
+1. Proposals and discussions
+    1. Announcement: Proposal update in wider CG
+    1. Discussion: Prototyping and investigation plans (10 - 15 minutes)
+    1. Presentation: dart2wasm (Aske Simon Christenson, 20-30 minutes)
+1. Closure
+
+## Meeting Notes
+
+### Introduction of attendees
+
+- Thomas Lively
+- Aske Simon Christensen
+- Ben Titzer
+- Zhi An Ng
+- Roberto Lublinerman
+- Rick Battagline
+- Ross Tate
+- Zalim Bashorov
+- Lars Hansen
+- Keith Miller
+- Tim Steenvoorden
+- Alon Zakai
+- Jakob Kummerow
+- Andreas Rossberg
+- Conrad Watt
+- Goktug Gokdogan
+- Luke Wagner
+- Francis McCabe
+- Mark (??)
+- Asumu Takikawa
+- Slaval Kuzmich
+
+###  Announcement: Proposal update in wider CG
+
+TL: Let me know if you want to join me, Alon, and Jakob in presenting recent progress on the GC proposal.
+
+### Presentation: dart2wasm (Aske Simon Christenson, 20-30 minutes)
+
+TODO: slides
+
+ASC: Dart is very dynamic. Fields can be overridden by getters and setters or vice versa. Even adding two integers is a call in theory. We depend on type flow analysis heavily for devirtualization in the front end. Every object has a class ID. Every method has a selector offset into the table. Call dispatch uses Wasm tables and call_indirect. Any class can be used as an interface, so all calls are interface calls?
+
+BT: you have reference types in func types, how closely do they need to match for call_indirect?
+
+ASC: I will get back to that, but parameters can change types in overrides. So the wasm signature needs to be the least upper bound. More details later.
+
+ZB (chat): Why do you have tree shaking 2 times?
+
+SK (chat): Does it mean that table size is (num_interface_methods * num_ClassIds)?
+
+ASC: no, these are interleaved. If you have the complete square table it will be very sparse. They are packed together. Each row corresponding to a particular selector will have entries for all the class ids that have that particular method, rows will be sparsed, and interleaved into a 1 dimensional table, so it's usually not very big
+
+FM: on this slide, do you mean each dart object is translated to a Wasm struct rather than each class?
+
+ASC: each class translates to a struct definition
+
+RT: What is the `context` field in _Function?
+
+ASC: It’s a pointer to the context.
+
+RT: Could you have the context be fields after the `functionRef` in e.g. _Function0?
+
+ASC: If you have two different functions with different contexts and different variables, they need to have the same type so the functions are interchangeable.
+
+RT: The two different functions being subclasses of _Function0 would accomplish that.
+
+ASC: There needs to be a cast anyway since the contexts can all be different, so I didn’t do that.
+
+AR: like flattened closures, save 1 indirection and allocation per closure
+
+ASC: you would put function ref in the same object as the context
```

Comment:
Post-meeting addendum: Multiple functions inside the same function can share a context. Mutations from one function are seen by the others. Therefore, there needs to be an indirection from the closure to the context (i.e. deep closures).